Support meta tags in StaticHTMLRenderer

AndrewBarba opened the pull request:
This PR adds the ability to control `<head>` tags using a new `HTMLTitle` view and `HTMLMeta` view. Taking inspiration from Next.js `<NextHead>`, you can use these views anywhere in your view hierarchy and they will be hoisted to the top `<head>` section of the html.

Use as a view:

```swift
var body: some View {
  VStack {
    ...
    HTMLTitle("Hello, Tokamak")
    HTMLMeta(charset: "utf-8")
    ...
  }
}
```

Use as a view modifier:

```swift
var body: some View {
  VStack {
    ...
  }
  .htmlTitle("Hello, Tokamak")
  .htmlMeta(charset: "utf-8")
}
```

And the resulting html (no matter where these are used in your view hierarchy):

```html
<html>
  <head>
    <title>Hello, Tokamak</title>
    <meta charset="utf-8">
  </head>
  <body>
    ...
  </body>
</html>
```

carson-katri commented:
There was previously some discussion in #269  around designing a way to modify the head through a Scene similar to the `Settings` scene in SwiftUI:

```swift
struct MyApp : App {
  var body: some Scene {
    Head {
      Meta(…)
    }
    WindowGroup(…) {
      ContentView()
    }
  }
}
```

MaxDesiatov commented:
@AndrewBarba would do you think about the suggested `Head`/`Meta` API that uses types conforming to `Scene`? I also think that we should support the `property` attribute for OpenGraph meta tags.

MaxDesiatov commented:
Yeah, that would require bigger changes to the renderer though. Let me know if you'd like to implement them yourself or have any questions. I can pick this up next week otherwise.

AndrewBarba commented:
This kind of makes sense. I would expect a new initializer that takes in a Scene which defines the entire HTML structure - including Meta tags and titles? And then the existing init that just takes in a View would wrap that View with this new base HTML structure. I have so confusion around `HTMLTarget`, if I make a full htmlTree as the target then all the children seem to be overwritten. Im not sure what the main use of the `HTMLTarget` struct actually is to be honest.

MaxDesiatov commented:
> I would expect a new initializer that takes in a Scene which defines the entire HTML structure - including Meta tags and titles?

No need for a new initializer, I don't think one would pass `Scene` values instead of `App` values, and already have an initializer for the latter.

> I have some confusion around `HTMLTarget`, if I make a full htmlTree as the target then all the children seem to be overwritten. Im not sure what the main use of the `HTMLTarget` struct actually is to be honest.

Sorry about the confusion. `HTMLTarget` is a class that represents tree state the reconciler updates when needed. I misremembered the logic here, you're right, initial `HTMLTarget` value will always be overwritten. You'll have to modify `extension WindowGroup: SceneDeferredToRenderer` to wrap `view` in `<head>` and `<body>` tags instead.

AndrewBarba commented:
@MaxDesiatov @carson-katri Check it out, I refactored PR to use preference keys so now you can drop `Title` and `Meta` anywhere in your view hierarchy just like next.js and its hoisted to the top. I'm pretty happy with end result but let me know if you want me to add scripts, styles or anything else. Also very open to feedback on how I organized the files.

```swift
var body: some View {
  VStack {
    ...
    Title("Hello, Tokamak")
    Meta(charset: "utf-8")
    ...
  }
}
```

You can see the result working here: https://tokamak.app.swift.cloud

AndrewBarba commented:
I agree, I think prefixing makes more sense. I made this change and updated the PR description to show the correct usage. I added usage for both views and view modifiers. SwiftUI on iOS uses a lot of modifiers, ex. `navigationBarTitle`, so I think it's important to support both.

MaxDesiatov commented on Sources/TokamakCore/StackReconciler.swift:
```diff
@@ -58,6 +58,12 @@ public final class StackReconciler<R: Renderer> {
    */
   public let rootTarget: R.TargetType
 
+  /** A root renderer's main preference store
```
nit: comment formatting 
```suggestion
  /** A root renderer's main preference store.
```

MaxDesiatov commented on Sources/TokamakStaticHTML/Preferences/Preferences.swift:
```diff
@@ -0,0 +1,41 @@
+
+import TokamakCore
+
+internal struct HTMLTitlePreferenceKey: PreferenceKey {
```
We don't use `internal` in our codebase as it's the default visibility and is redundant.

```suggestion
struct HTMLTitlePreferenceKey: PreferenceKey {
```

MaxDesiatov reviewed (commented):
There are a few nits to pick, but overall this makes sense. One more question: how does this behave if there are multiple `Title` views in the same tree? Whichever comes last is applied I guess? Would you be able to add a snapshot test that confirms this? Thanks!

carson-katri reviewed (changes requested):
Overall it looks good to me, just needs a couple tests. I think they'd go in `TokamakStaticHTMLTests/HTMLTests.swift`

carson-katri commented:
Could you add a test for `HTMLMeta` too?

AndrewBarba commented:
Definitely - working on them now

AndrewBarba commented:
I pushed a test that uses 4 meta tags but only the last one ends up in the resulting html. Currently trying to debug this, I'm pretty sure my implementation is correct so I wonder if there's an issue with the preference store.

AndrewBarba commented:
Okay the main issue here is that every mount target gets its own preference store and then we are trying to merge upwards anytime a child store changes. This is problematic because the merge only saves the value of the child and doesn't maintain any values in the parent for the same key.

I was able to fix this by:

1. Change `_PreferenceStore` to a class
2. Initialize `MountedElement.preferenceStore` to the parents store or create a new one

I can push this up if you think it makes sense, otherwise I'm open to other approaches.

MaxDesiatov commented:
IIRC @carson-katri is the author of preference store code, so I'll defer to him on this matter if he has a moment to review this.

carson-katri commented:
So does this make the child take all the values of the parent's preference store? If so, I don't think this behavior is correct as preference values flow up the tree not down. Similar to how an environment value from a child is not set on the parent, a preference value from a parent is not set on the child.

AndrewBarba commented:
Ah yes you are right. Don't think we have tests for this so I will add them. 

So I think the approach to solve that is add a weak parent store to the _PreferenceStore. Each mount target will go back to getting its own store but the pref store will be updated to modify its parent store anytime one of its keys changes. This is still a better solution to merge, because the issue with merge is it doesn't know which keys changed, and so even with some fancy generic work to deep merge keys upwards, it was over adding values to the parent store because it was doing a full merge on ever didSet. If we control the upstream changes to only happen on inserting a key this should be more reliable.

AndrewBarba commented:
Okay just pushed a test that I think correctly covers what you described.

carson-katri reviewed (approved):
That looks good to me, thanks!

MaxDesiatov reviewed (changes requested):
Overall this is great, one small change request: could you run SwiftFormat, and potentially have a pre-commit hook installed as described in [the "Coding Style" section of `CONTRIBUTING.md`](https://github.com/TokamakUI/Tokamak/blob/main/CONTRIBUTING.md#coding-style)? Some of the newly added files use 4 spaces for indentation, while we use 2 spaces everywhere in the project.

Thanks!

AndrewBarba commented:
Yes definitely. Xcode is so annoying with the indents, I thought I caught them all. Will get this fixed

MaxDesiatov commented:
Don't worry about catching them in Xcode manually. When I edit in Xcode I just write some messy inconsistent indentation, and it's all reformatted with pre-commit on every commit automatically 🙂

AndrewBarba commented:
Yeah wow - this is so much better. My bad for not reading that code style doc earlier

MaxDesiatov reviewed (approved):
Lovely, thanks!
